runtime/src/iree/hal/cts/driver_test.h

```diff
@@ -22,25 +22,124 @@ namespace cts {
 
-  std::cout << "Driver has " << device_info_count << " device(s)";
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
   for (iree_host_size_t i = 0; i < device_info_count; ++i) {
     std::cout << "  Creating device '"
               << std::string(device_infos[i].name.data,
                              device_infos[i].name.size)
-              << "'";
+              << "'..\n";
     iree_hal_device_t* device = NULL;
     IREE_ASSERT_OK(iree_hal_driver_create_device_by_id(
         driver_, device_infos[i].device_id, /*param_count=*/0, /*params=*/NULL,
         iree_allocator_system(), &device));
     iree_string_view_t device_id = iree_hal_device_id(device);
-    std::cout << "  Created device with id: '"
-              << std::string(device_id.data, device_id.size) << "'";
+    std::cout << "    Created device with id: '"
+              << std::string(device_id.data, device_id.size) << "'\n";
+    iree_hal_device_release(device);
+  }
+
+  iree_allocator_free(iree_allocator_system(), device_infos);
+}
+
+TEST_P(driver_test, QueryAndCreateAvailableDevicesByOrdinal) {
+  iree_host_size_t device_info_count = 0;
+  iree_hal_device_info_t* device_infos = NULL;
+  IREE_ASSERT_OK(iree_hal_driver_query_available_devices(
+      driver_, iree_allocator_system(), &device_info_count, &device_infos));
+
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
+  for (iree_host_size_t i = 0; i < device_info_count; ++i) {
+    std::cout << "  Creating device '"
+              << std::string(device_infos[i].name.data,
+                             device_infos[i].name.size)
+              << "'..\n";
+    iree_hal_device_t* device = NULL;
+    IREE_ASSERT_OK(iree_hal_driver_create_device_by_ordinal(
+        driver_, i, /*param_count=*/0, /*params=*/NULL, iree_allocator_system(),
+        &device));
+    iree_string_view_t device_id = iree_hal_device_id(device);
+    std::cout << "    Created device with id: '"
+              << std::string(device_id.data, device_id.size) << "'\n";
+    iree_hal_device_release(device);
+  }
+
+  iree_allocator_free(iree_allocator_system(), device_infos);
+}
+
+TEST_P(driver_test, QueryAndCreateAvailableDevicesByPath) {
+  iree_host_size_t device_info_count = 0;
+  iree_hal_device_info_t* device_infos = NULL;
+  IREE_ASSERT_OK(iree_hal_driver_query_available_devices(
+      driver_, iree_allocator_system(), &device_info_count, &device_infos));
+
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
+
+  if (device_info_count > 0) {
+    iree_string_view_t name = device_infos[0].name;
+    std::cout << "  Creating device '" << std::string(name.data, name.size)
+              << "' with empty path..\n";
+    iree_hal_device_t* device = NULL;
+    IREE_ASSERT_OK(iree_hal_driver_create_device_by_path(
+        driver_, name, iree_string_view_empty(),
+        /*param_count=*/0, /*params=*/NULL, iree_allocator_system(), &device));
```

Comment:
This method is not yet implemented for all drivers:
https://github.com/iree-org/iree/actions/runs/4098806533/jobs/7068179944
```
Expected: error code OK
  Actual: 0x260b725, whose error code is NOT_FOUND: /work/runtime/src/iree/hal/drivers/local_sync/sync_driver.c:143: NOT_FOUND; device paths not yet implemented
```

Reply:
Yeah, local sync/task is missing implementation for it. I've implemented them in c648d33.